UI/Sound.cpp

```diff
@@ -144,31 +156,221 @@ Sound& Sound::GetSound()
 }
 
 Sound::Sound() :
+    pimpl(new SoundImpl)
+{}
+
+// Require here because SoundImpl is defined in this file.
+Sound::~Sound() {}
+
+void Sound::Enable()
+{ pimpl->Enable(); }
+
+void Sound::Disable()
+{ pimpl->Disable(); }
+
+void Sound::PlayMusic(const boost::filesystem::path& path, int loops /* = 0*/)
+{ pimpl->PlayMusic(path, loops); }
+
+void Sound::StopMusic()
+{ pimpl->StopMusic(); }
+
+void Sound::PlaySound(const boost::filesystem::path& path, bool is_ui_sound/* = false*/)
+{ pimpl->PlaySound(path, is_ui_sound); }
+
+void Sound::FreeSound(const boost::filesystem::path& path)
+{ pimpl->FreeSound(path); }
+
+void Sound::FreeAllSounds()
+{ pimpl->FreeAllSounds(); }
+
+void Sound::SetMusicVolume(int vol)
+{ pimpl->SetMusicVolume(vol); }
+
+void Sound::SetUISoundsVolume(int vol)
+{ pimpl->SetUISoundsVolume(vol); }
+
+void Sound::DoFrame()
+{ pimpl->DoFrame(); }
+
+
+
+////////////////////////////////////////////////////////////
+// SoundImpl
+////////////////////////////////////////////////////////////
+
+Sound::SoundImpl::SoundImpl() :
     m_sources(),
-    m_music_loops(),
+    m_music_loops(0),
     m_music_name(),
     m_buffers(),
     m_music_buffers(),
     m_ogg_file(),
     m_ogg_format(),
     m_ogg_freq(),
-    m_temporary_disable_count(0)
+    m_temporary_disable_count(0),
+    m_initialized(false)
 {
-    InitOpenAL(NUM_SOURCES, m_sources, m_music_buffers);
+    if (!GetOptionsDB().Get<bool>("UI.sound.enabled") && !GetOptionsDB().Get<bool>("UI.sound.music-enabled"))
+        return;
+
+    try {
+        Enable();
+    } catch (std::runtime_error const & e) {
+        if (std::string(e.what()) != "ERROR_SOUND_INITIALIZATION_FAILED")
+            throw;
+    }
+}
+
+Sound::SoundImpl::~SoundImpl() {
+    Disable();
 }
 
-Sound::~Sound() {
+void Sound::SoundImpl::Enable() {
+    if (m_initialized)
+        return;
+
+    //Reset playing audio
+    m_music_loops = 0;
+    m_buffers.clear();
+    m_temporary_disable_count = 0;
+
+    InitOpenAL();
+    if (!m_initialized) {
+        ErrorLogger() "Unable to initialize audio.  Sound effects and music are disabled.";
+        // TODO: Let InitOpenAL throw the OpenAL error message which
+        // might be more useful.
+        throw std::runtime_error("ERROR_SOUND_INITIALIZATION_FAILED");
```

Comment:
should probably make a derived exception class rather than testing the value of the exception text...